### PR TITLE
Siddhi + Analytics extension archetypes

### PR DESCRIPTION
**Pull request by hmrajas**

### CODE REVIEW

File: analytics-event-publisher-archetype/pom.xml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <artifactId>carbon-extension-archetype</artifactId>
+        <groupId>org.wso2.carbon.extension.archetype</groupId>
+        <version>1.0.2-SNAPSHOT</version>
+    </parent>
+
+    <packaging>maven-archetype</packaging>
+    <artifactId>org.wso2.carbon.extension.archetype.analytics.publisher-archetype</artifactId>
```
Comment (mohanvive) [inline analytics-event-publisher-archetype/pom.xml]: Here, the word archetype is there in two places of the artifactId. Do we really need that ?

File: siddhi-aggregator-archetype/pom.xml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <artifactId>carbon-extension-archetype</artifactId>
+        <groupId>org.wso2.carbon.extension.archetype</groupId>
+        <version>1.0.2-SNAPSHOT</version>
+    </parent>
+
+    <groupId>org.wso2.carbon.extension.archetype</groupId>
+    <packaging>maven-archetype</packaging>
+    <artifactId>org.wso2.carbon.extension.archetype.siddhi.aggregator-archetype</artifactId>
```
Comment (mohanvive) [inline siddhi-aggregator-archetype/pom.xml]: We cannot have carbon here ? Siddhi is not a carbon component.. Why we are having carbon in artifactId ?

File: siddhi-stream-processor-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_processor_name__StreamProcessor.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * ${siddhi_stream_processor_name}
+ */
+
+public class ${siddhi_stream_processor_name}StreamProcessor extends StreamProcessor {
+
```
Comment (mohanvive) [inline siddhi-stream-processor-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_processor_name__StreamProcessor.java]: @hmrajas , In the review we have asked you to write a specific scenario for streamProcessor Isn't ? Can you implement the stream processor  logic as similar to Time Window and include the event expiry time as additional attribute please ? That will be very clear..

File: esb-connector-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml [hunk 4]
```diff
@@ -1,38 +1,55 @@
...
-        <include>**/*.java</include>
-      </includes>
-    </fileSet>
-  </fileSets>
-  <requiredProperties>
-      
-      <requiredProperty key="connector_name"/>
-      
+<!--
+ ~ Copyright (c) 2015 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (kesavany) [inline esb-connector-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml]: Change the year in licence header.

Comment (hmrajas) [inline esb-connector-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml]: Already fixed in https://github.com/wso2-extensions/archetypes/pull/8

File: siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    /**
+     * This will be called only once and this can be used to acquire
+     * required resources for the processing element.
+     * This will be called after initializing the system and before
+     * starting to process the events.
+     */
+    @Override
+    public void start() {
+        //Implement start logic to acquire relevant resources
```
Comment (kesavany) [inline siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java]: Require full stop in ending main comment lines.

File: siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+    @Override
+    public Attribute.Type getReturnType() {
+        return Attribute.Type.DOUBLE;
+    }
+
+
+    /**
+     * Used to collect the serializable state of the processing element, that need to be
+     * persisted for the reconstructing the element to the same state on a different point of time
```
Comment (kesavany) [inline siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java]: Require full stop in ending main comment line(class level, methods etc..). Fix in all places.

File: siddhi-stream-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_function_name__StreamFunctionProcessor.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+ * ${siddhi_stream_function_name} custom streamFunctionProcessor
+ */
+public class ${siddhi_stream_function_name}StreamFunctionProcessor extends StreamFunctionProcessor {
+
+    private static final Logger LOGGER = Logger.getLogger(${siddhi_stream_function_name}StreamFunctionProcessor.class);
+    private final Geocoder geocoder = new Geocoder();
+    private boolean debugModeOn;
+
+    /**
+     * The process method of the ${siddhi_stream_function_name}StreamFunctionProcessor,
```
Comment (kesavany) [inline siddhi-stream-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_function_name__StreamFunctionProcessor.java]: Require full stop in ending main comment line (class level, methods etc..).

File: siddhi-stream-processor-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_processor_name__StreamProcessor.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+                        expiredEventChunk.remove();
+                        expiredEvent.setTimestamp(currentTime);
+                        streamEventChunk.insertBeforeCurrent(expiredEvent);
+                    } else {
+                        break;
+                    }
+                }
+
+                if (streamEvent.getType() == StreamEvent.Type.CURRENT) {
+
```
Comment (kesavany) [inline siddhi-stream-processor-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_processor_name__StreamProcessor.java]: Remove the empty line.

File: analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/test/java/__carbon_analytics_publisher_name__TestCase.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                loggedInSessionCookie);
+        eventStreamManagerAdminServiceClient = configurationUtil.getEventStreamManagerAdminServiceClient(backendURL,
+                loggedInSessionCookie);
+        eventPublisherAdminServiceClient = configurationUtil.getEventPublisherAdminServiceClient(backendURL,
+                loggedInSessionCookie);
+    }
+
+    @Test(groups = {"wso2.cep"}, description = "Testing ${carbon_analytics_publisher_name} publisher ")
+    public void TestScenario() throws Exception {
+
```
Comment (kesavany) [inline analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/test/java/__carbon_analytics_publisher_name__TestCase.java]: Remove the empty line.

File: analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/main/java/__carbon_analytics_publisher_name__EventAdapter.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Publish events. Throws ConnectionUnavailableException if it cannot connect to the back end.
+     *
+     * @param o
+     * @param map
+     * @throws ConnectionUnavailableException
+     */
+    public void publish(Object o, Map<String, String> map) throws ConnectionUnavailableException {
+
```
Comment (kesavany) [inline analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/main/java/__carbon_analytics_publisher_name__EventAdapter.java]: Remove the empty line.Fix in all places.

File: siddhi-aggregator-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_aggregator_name__AggregateFunction.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+    /**
+     * The initialization method for ${siddhi_aggregator_name}AggregatorFunction
+     *
+     * @param attributeExpressionExecutors are the executors of each attributes in the function
+     * @param executionPlanContext         Execution plan runtime context
+     */
+    @Override
+    protected void init(ExpressionExecutor[] attributeExpressionExecutors, ExecutionPlanContext executionPlanContext) {
+
```
Comment (kesavany) [inline siddhi-aggregator-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_aggregator_name__AggregateFunction.java]: Remove the empty line.

File: siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java (hunk 5 above)
Comment (hmrajas) [inline siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java]: Fixed in https://github.com/wso2-extensions/archetypes/pull/10

File: siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java (hunk 6 above)
Comment (hmrajas) [inline siddhi-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_function_name__Function.java]: Fixed in #10

File: siddhi-stream-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_function_name__StreamFunctionProcessor.java (hunk 7 above)
Comment (hmrajas) [inline siddhi-stream-function-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_function_name__StreamFunctionProcessor.java]: Fixed in #10

File: siddhi-stream-processor-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_processor_name__StreamProcessor.java (hunk 8 above)
Comment (hmrajas) [inline siddhi-stream-processor-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_stream_processor_name__StreamProcessor.java]: Fixed in #10

File: analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/test/java/__carbon_analytics_publisher_name__TestCase.java (hunk 9 above)
Comment (hmrajas) [inline analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/test/java/__carbon_analytics_publisher_name__TestCase.java]: Fixed in #10

File: analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/main/java/__carbon_analytics_publisher_name__EventAdapter.java (hunk 10 above)
Comment (hmrajas) [inline analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/main/java/__carbon_analytics_publisher_name__EventAdapter.java]: Fixed in #10

File: siddhi-aggregator-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_aggregator_name__AggregateFunction.java (hunk 11 above)
Comment (hmrajas) [inline siddhi-aggregator-archetype/src/main/resources/archetype-resources/src/main/java/__siddhi_aggregator_name__AggregateFunction.java]: Fixed in #10

File: analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/main/java/__carbon_analytics_publisher_name__EventAdapterFactory.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /**
+     * Specify supported message formats for the created publisher type.
+     *
+     * @return
+     */
+    @Override
+    public List<String> getSupportedMessageFormats() {
+        return null;
```
Comment (pamod) [inline analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/main/java/__carbon_analytics_publisher_name__EventAdapterFactory.java]: Any reason for returning null ? should it be an empty collection instead ? also would it make it more applicable to indicate NotImplimentedException instead ?

Comment (hmrajas) [inline analytics-event-publisher-archetype/src/main/resources/archetype-resources/src/main/java/__carbon_analytics_publisher_name__EventAdapterFactory.java]: There is no special reason to leave it as null.
Fixed in https://github.com/wso2-extensions/archetypes/pull/10

**Pull request merged**